src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -1541,72 +1267,342 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     return mostRecentCurrentTime;
   }
 
+  function deferredUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = computeAsyncExpiration();
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  function syncUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = Sync;
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  // TODO: Everything below this is written as if it has been lifted to the
+  // renderers. I'll do this in a follow-up.
```

Comment:
I think the easiest way to review this would be to start here. Everything after this line is meant to be extracted into the renderer. It's all logic that already existed, but used to be spread out across the scheduler.